include/aspect/fe_variable_system.h

```diff
@@ -0,0 +1,423 @@
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef __aspect__fe_variable_system_h
+#define __aspect__fe_variable_system_h
+
+#include <aspect/global.h>
+
+namespace dealii
+{
+  /**
+   * Structure to describe everything necessary to define a single variable of
+   * the class FEVariableSystem. It groups the FiniteElement<dim> with other
+   * information like a name and the block structure used in linear systems. A
+   * vector of these structs will be converted into an ordered collection of
+   * objects of type an FEVariable represented by FEVariableSystem.
+   *
+   * Blocks are used to extract or solve for certain variables separately or
+   * in block based preconditioners.  As an example, in a Stokes problem one
+   * might want velocity and pressure in a single block (for use with a direct
+   * solver), or one block for all velocity components and a second block for
+   * the pressure (for a Schur complement-based preconditioner), or separate
+   * blocks for pressure and each velocity component.
+   */
+  template <int dim>
+  struct VariableDefinition
+  {
+    /**
+     * Constructor.
+     *
+     * @param name A user-friendly and unique string representation of the variable.
+     *
+     * @param fe The FiniteElement class to use.
+     *
+     * @param multiplicity Number of copies of the @p fe to create.
+     *
+     * @param n_blocks Number of blocks this variable represents inside the
+     * linear system. A value of 0 will add the next variable to the current
+     * block, while 1 will put the next variable into a new block. A value
+     * that is equal to the number of components will create a block for each
+     * component.
+     */
+    VariableDefinition(const std::string &name,
+                       std_cxx11::shared_ptr<FiniteElement<dim> > fe,
+                       const unsigned int multiplicity,
+                       const unsigned int n_blocks);
+
+    /**
+     * Return the total number of components of this variable.
+     */
+    unsigned int n_components() const;
+
+    /**
+     * Name of the variable used in FEVariableSystem to identify it.
+     */
+    std::string name;
+    /**
+     * The FiniteElement space.
+     */
+    std_cxx11::shared_ptr<FiniteElement<dim> > fe;
+    /**
+     * The multiplicity used in FESystem: how many copies of @p fe are there?
+     */
+    unsigned int multiplicity;
+    /**
+     * Number of linear algebra blocks occupied by this variable. See
+     * constructor for details.
+     */
+    unsigned int n_blocks;
+  };
+
+  /**
+   * Struct to represent a variable as part of a FEVariableSystem. Constructed
+   * from an instance of VariableDefinition<dim> but contains additional
+   * information that can be queried.
+   */
+  template<int dim>
+  struct FEVariable: public VariableDefinition<dim>
```

Comment:
Hm, you're right. "Element" (as in "element of a collection") is also bad.

I didn't like the name `FEVariable` because the name does not make clear any kind of distinction between what a `FEVariable` and a `VariableDefinition` is. They sound the same to me. The difference, semantically, is that the latter stands on its own, whereas the former is an element in a collection. So maybe `FECollectionComponent` or `FECollectionElement`, and then name the next class `FECollection`? Ir `VariableCollectionComponent` and `VariableCollection` (dropping the `FE` altogether)?